objects/src/mock/chain.rs

```diff
@@ -55,3 +55,401 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+
+        /// Storage for the historical state of the chain.
+        ///
+        /// This store contains historical data for accounts, nullifiers, notes, and batches.
+        data: DefaultMerkleStore,
+
+        /// History of produced blocks.
+        blocks: Vec<BlockHeader>,
+
+        /// Tree containing the latest `Nullifier`'s tree.
+        nullifiers: TieredSmt,
+
+        /// Tree containing the latest hash of each account.
+        accounts: SimpleSmt,
+
+        /// RNG used to seed builders.
+        ///
+        /// This is used to seed the [AccountBuilder] and the [NonFungibleAssetBuilder].
+        rng: R,
+
+        /// Builder for new [AccountId]s of faucets.
+        account_id_builder: AccountIdBuilder<R>,
+
+        /// Objects that have been created and commited to a block.
+        ///
+        /// These can be used to perform additional operations on a block.
+        ///
+        /// Note:
+        /// - The [Note]s in this container have the `proof` set.
+        objects: Objects<R>,
+
+        /// Objects that have been created and are waiting for a block.
+        ///
+        /// These objects will become available once the block is sealed.
+        ///
+        /// Note:
+        /// - The [Note]s in this container do not have the `proof` set.
+        #[cfg_attr(feature = "serde", serde(skip))]
+        pending_objects: Objects<R>,
+    }
+
+    #[derive(Debug, PartialEq, Eq)]
+    pub enum OnChain {
+        No,
+        Yes,
+    }
+
+    #[derive(Debug, PartialEq, Eq)]
+    pub enum Immutable {
+        No,
+        Yes,
+    }
+
+    pub enum Error {
+        DuplicatedNullifier,
+        DuplicatedNote,
+    }
+
+    impl<R: Rng + SeedableRng> MockChain<R> {
+        pub fn new(mut rng: R) -> Result<Self, rand::Error> {
+            let account_rng = R::from_rng(&mut rng).expect("rng seeding failed");
+            let account_id_builder = AccountIdBuilder::new(account_rng);
+            Ok(Self {
+                chain: ChainMmr::default(),
+                data: DefaultMerkleStore::default(),
+                blocks: vec![],
+                nullifiers: TieredSmt::default(),
+                accounts: SimpleSmt::new(64).expect("depth too big for SimpleSmt"),
+                rng,
+                account_id_builder,
+                objects: Objects::new(),
+                pending_objects: Objects::new(),
+            })
+        }
+
+        /// Creates the next block.
+        ///
+        /// This will also make all the objects currently pending available for use.
+        pub fn seal_block(&mut self) -> Result<BlockHeader, MerkleError> {
+            let block_num: u64 = self.blocks.len().try_into().expect("usize to u64 failed");
+            let block_num: Felt = block_num.into();
+
+            for account in self.pending_objects.accounts.iter() {
+                let id: Felt = account.id().into();
+                self.accounts.update_leaf(id.as_int(), account.hash().into())?;
+            }
```

Comment:
Note: This is building on top of the tree with the account state from the previous blocks.